### PR TITLE
[Easy] Use ConstantSource for partials kwargs

### PR DESCRIPTION
**Pull request by voznesenskym**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #110639

[Easy] Eagerly propagate guards for functools partial

cc @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @chenyang78 @aakhundov @kadeng

### CODE REVIEW

**Comment by ezyang** [comment]
I am not opposed to this as a workaround but the implementation is not tight. In particular, it means any functools.partial in scope gets guarded on, even if it is completely irrelevant for compilation.

Evidence: this modified test script

```
import torch
from functools import partial

def bar(x, **kwargs):
    return x + x

@torch.compile(backend="eager", dynamic=True)
def foo(x, i):
    def inner():
        print("this is a graph_break")
        return x * 2
        return op(x)

    op = partial(bar, dim=10)
    x = inner()
    op = partial(bar, other=10)
    return inner() + x

foo(torch.rand(1), 10)
```

Produces these logs

https://gist.github.com/ezyang/89453c84d9b5f74ccc1608ae3c10c2a3#file-gistfile0-txt-L286

Notably you can see that despite op not actually being used at all (and merely being present in free vars) is sufficient to cause guarding on the structure of functools.partial

**Review by anijain2305 — APPROVED** [review]

**Comment by voznesenskym** [comment]
> I am not opposed to this as a workaround but the implementation is not tight. In particular, it means any functools.partial in scope gets guarded on, even if it is completely irrelevant for compilation.
> 
> Evidence: this modified test script
> 
> ```
> import torch
> from functools import partial
> 
> def bar(x, **kwargs):
>     return x + x
> 
> @torch.compile(backend="eager", dynamic=True)
> def foo(x, i):
>     def inner():
>         print("this is a graph_break")
>         return x * 2
>         return op(x)
> 
>     op = partial(bar, dim=10)
>     x = inner()
>     op = partial(bar, other=10)
>     return inner() + x
> 
> foo(torch.rand(1), 10)
> ```
> 
> Produces these logs
> 
> https://gist.github.com/ezyang/89453c84d9b5f74ccc1608ae3c10c2a3#file-gistfile0-txt-L286
> 
> Notably you can see that despite op not actually being used at all (and merely being present in free vars) is sufficient to cause guarding on the structure of functools.partial

I agree - this is an aggressive overspecialization. It is not tight. It's meant to unblock @Fidget-Spinner 

In the original case, the `keywords.dim` was being guarded on elsewhere.

I am working on an alternate fix that removes that from the guards instead, which might suffice to fix this.

**Comment by voznesenskym** [comment]
> > I am not opposed to this as a workaround but the implementation is not tight. In particular, it means any functools.partial in scope gets guarded on, even if it is completely irrelevant for compilation.
> > Evidence: this modified test script
> > ```
> > import torch
> > from functools import partial
> > 
> > def bar(x, **kwargs):
> >     return x + x
> > 
> > @torch.compile(backend="eager", dynamic=True)
> > def foo(x, i):
> >     def inner():
> >         print("this is a graph_break")
> >         return x * 2
> >         return op(x)
> > 
> >     op = partial(bar, dim=10)
> >     x = inner()
> >     op = partial(bar, other=10)
> >     return inner() + x
> > 
> > foo(torch.rand(1), 10)
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Produces these logs
> > https://gist.github.com/ezyang/89453c84d9b5f74ccc1608ae3c10c2a3#file-gistfile0-txt-L286
> > Notably you can see that despite op not actually being used at all (and merely being present in free vars) is sufficient to cause guarding on the structure of functools.partial
> 
> I agree - this is an aggressive overspecialization. It is not tight. It's meant to unblock @Fidget-Spinner
> 
> In the original case, the `keywords.dim` was being guarded on elsewhere.
> 
> I am working on an alternate fix that removes that from the guards instead, which might suffice to fix this.

The root case is actually:

```
if not isinstance(self.get_source(), RandomValueSource):
                guards = {self.get_source().make_guard(GuardBuilder.TYPE_MATCH, True)}
                options = {"guards": guards}
```

This can be fixed with a ConstantSource I think

**Review by Fidget-Spinner — APPROVED** [review]
LGTM. `functools.partial.kwargs` should be immutable in dynamo anyways (IIRC, we fail if the user tries to modify it manually)

**Comment by voznesenskym** [comment]
> LGTM. `functools.partial.kwargs` should be immutable in dynamo anyways (IIRC, we fail if the user tries to modify it manually)

I think it doesn't work in certain cases. I am going to go back to the drawing board.

**Comment by voznesenskym** [comment]
The real bug here is ints promoted to arguments where their source has attr access. In this case we promote dim aka 10 to an input, but, its source is op.keywords['dim']. We need to walk the chain of sources and accumulate attribute access guards anytime we lift something - we cannot naively use its source, thinking it is safe.

Will think about this a bit and make a new PR. @Fidget-Spinner if you are blocked, feel free to take the first version of this PR that over aggressively works around the case you care about via `self.tx.output.guards.update(guards)` 
.
